diff --git a/src/x11/meta-x11-display.c b/src/x11/meta-x11-display.c
index 1b3e792..a4061fb 100644
--- a/src/x11/meta-x11-display.c
+++ b/src/x11/meta-x11-display.c
@@ -1329,6 +1329,9 @@ meta_x11_display_init_frames_client (MetaX11Display *x11_display)
   x11_display->frames_client_cancellable = g_cancellable_new ();
   x11_display->frames_client = meta_frame_launch_client (x11_display,
                                                          display_name);
+  if (!x11_display->frames_client)
+    return;
+
   g_subprocess_wait_async (x11_display->frames_client,
                            x11_display->frames_client_cancellable,
                            on_frames_client_died, x11_display);
